File: clients/sdk-cookbook/src/kit/segments.ts
/**
 * Segment results and the expect-block contract.
 *
 * The journey runs every segment for real. A segment that today cannot pass
 * because of a named open issue carries an `expectBlock`. The block never
 * weakens the assertion and never stubs the segment out: the segment still
 * runs, still asserts the spec-correct behavior, and its real failure text is
 * kept in the report.
 *
 * The contract has two directions, and the second one is the point:
 *
 *  - assertion fails with the SIGNATURE the named issues cause ->
 *    `expectBlocked`. The journey stays green and prints the issue numbers.
 *  - assertion fails with anything else -> `failed`. A block excuses exactly
 *    the defect it names and nothing else, so an unrelated regression in a
 *    blocked segment still reds the journey.
 *  - assertion PASSES while an expect-block is declared -> `unblocked`.
 *    The journey FAILS. The fix landed, so the segment must be promoted to
 *    required by deleting its `expectBlock`. A stale block cannot rot
 *    silently, and the promotion happens the first time CI runs after the
 *    fix — that is the automatic light-up.
 */

/** What actually happened when a segment ran. */
export type SegmentOutcome =
  /** Ran and asserted clean. Required for the journey to be green. */
  | "passed"
  /** Failed exactly as its named open issues predict. Not fatal. */
  | "expectBlocked"
  /** Passed while expect-blocked: the block is stale. FATAL. */
  | "unblocked"
  /** Failed with no expect-block, or in a way no block predicted. FATAL. */
  | "failed";

/** Why a segment cannot pass yet, in the issue tracker's own numbers. */
export interface ExpectBlock {
  /** Open issue numbers. Empty is never valid. */
  readonly issues: readonly number[];
  /** One plain sentence: what those issues do to this segment. */
  readonly because: string;
  /**
   * The failure text those issues produce. The block excuses ONLY a failure
   * matching this; anything else is a real failure. Without it a block would
   * quietly absorb every future regression in its segment.
   */
  readonly signature: RegExp;
}

/** One named step of the journey. */
export interface Segment<Context> {
  readonly id: string;
  /** Plain-words title, used verbatim in the printed report. */
  readonly title: string;
  /** Present only while a named open issue stops this segment passing. */
  readonly expectBlock?: ExpectBlock;
  /** Runs the real work and throws on any failed assertion. */
  run(context: Context): Promise<void>;
}

export interface SegmentResult {
  readonly id: string;
  readonly title: string;
  readonly outcome: SegmentOutcome;
  readonly durationMs: number;
  readonly expectBlock: ExpectBlock | undefined;
  /** The real failure text, kept whether or not the failure was expected. */
  readonly failure: string | undefined;
}

export interface JourneyReport {
  readonly segments: readonly SegmentResult[];
  /** True only when no segment is `failed` or `unblocked`. */
  readonly ok: boolean;
}

function describe(error: unknown): string {
  if (error instanceof Error) {
    const cause = error.cause === undefined ? "" : `\n  caused by: ${String(error.cause)}`;
    return `${error.message}${cause}`;
  }
  return String(error);
}

export function classify(
  expectBlock: ExpectBlock | undefined,
  failure: string | undefined,
): SegmentOutcome {
  if (expectBlock === undefined) return failure === undefined ? "passed" : "failed";
  if (failure === undefined) return "unblocked";
  return expectBlock.signature.test(failure) ? "expectBlocked" : "failed";
}

/**
 * Runs one segment and classifies it. `now` is injected so the deterministic
 * unit tests can drive duration without a clock.
 */
export async function runSegment<Context>(
  segment: Segment<Context>,
  context: Context,
  now: () => number = () => performance.now(),
): Promise<SegmentResult> {
  const started = now();
  let failure: string | undefined;
  try {
    await segment.run(context);
  } catch (error) {
    failure = describe(error);
  }
  return {
    id: segment.id,
    title: segment.title,
    outcome: classify(segment.expectBlock, failure),
    durationMs: Math.round(now() - started),
    expectBlock: segment.expectBlock,
    failure,
  };
}

/**
 * Run a journey's segments in order, ALWAYS run `teardown`, then summarize.
 *
 * This is the loop every recipe would otherwise hand-copy, and the copy's
 * risky part is the `finally`: the SDK has no kill path (#15943), so a
 * recipe that forgets it leaks a spawned host child on the first failing
 * segment (PR #24319 review). Teardown is a callback because the kit does
 * not know what a journey's context owns.
 */
export async function runJourney<Context>(
  segments: ReadonlyArray<Segment<Context>>,
  context: Context,
  teardown: (context: Context) => Promise<void>,
): Promise<JourneyReport> {
  const results: SegmentResult[] = [];
  try {
    for (const segment of segments) {
      results.push(await runSegment(segment, context));
    }
  } finally {
    await teardown(context);
  }
  return summarize(results);
}

export function summarize(segments: readonly SegmentResult[]): JourneyReport {
  return {
    segments,
    ok: !segments.some(
      (segment) => segment.outcome === "failed" || segment.outcome === "unblocked",
    ),
  };
}

const MARK: Record<SegmentOutcome, string> = {
  passed: "PASS ",
  expectBlocked: "BLOCK",
  unblocked: "STALE",
  failed: "FAIL ",
};

function issueList(block: ExpectBlock): string {
  return block.issues.map((issue) => `#${String(issue)}`).join(", ");
}

/** The human-facing report. Every expect-block states its issue numbers. */
export function formatReport(report: JourneyReport): string {
  const lines: string[] = [];
  for (const segment of report.segments) {
    lines.push(
      `${MARK[segment.outcome]} ${segment.id.padEnd(26)} ${String(segment.durationMs).padStart(6)}ms  ${segment.title}`,
    );
    const block = segment.expectBlock;
    if (block !== undefined && segment.outcome === "expectBlocked") {
      lines.push(`        expect-blocked on ${issueList(block)}: ${block.because}`);
      lines.push(`        observed: ${indent(segment.failure ?? "(no detail)")}`);
    }
    if (block !== undefined && segment.outcome === "unblocked") {
      lines.push(
        `        EXPECT-BLOCK IS STALE. This segment now passes, so ${issueList(block)} looks fixed.`,
      );
      lines.push(
        `        Promote it: delete this segment's expectBlock so the journey requires it from now on.`,
      );
    }
    if (segment.outcome === "failed") {
      if (block !== undefined) {
        lines.push(
          `        This failure does NOT match the expect-block signature for ${issueList(block)}` +
            ` (${String(block.signature)}), so it is a real failure.`,
        );
      }
      lines.push(`        ${indent(segment.failure ?? "(no detail)")}`);
    }
  }
  const counts = new Map<SegmentOutcome, number>();
  for (const segment of report.segments) {
    counts.set(segment.outcome, (counts.get(segment.outcome) ?? 0) + 1);
  }
  const tally = (["passed", "expectBlocked", "unblocked", "failed"] as const)
    .map((outcome) => `${outcome}=${String(counts.get(outcome) ?? 0)}`)
    .join(" ");
  lines.push(`journey ${report.ok ? "OK" : "NOT OK"} (${tally})`);
  return lines.join("\n");
}

function indent(text: string): string {
  return text.split("\n").join("\n        ");
}

File: python/clients/sdk-quickstart-py/src/quickstart_journey/segments.py
"""Segment results and the expect-block contract — the Python twin of
``clients/sdk-cookbook/src/kit/segments.ts`` (spec 638 FR-638-022 / T040).

The journey runs every segment for real. A segment that today cannot pass
because of a named open issue carries an ``expect_block``. The block never
weakens the assertion and never stubs the segment out: the segment still
runs, still asserts the spec-correct behavior, and its real failure text is
kept in the report.

The contract has two directions, and the second one is the point:

- assertion fails with the SIGNATURE the named issues cause →
  ``expectBlocked``. The journey stays green and prints the issue numbers.
- assertion fails with anything else → ``failed``. A block excuses exactly
  the defect it names and nothing else.
- assertion PASSES while an expect-block is declared → ``unblocked``. The
  journey FAILS: the fix landed, so the segment must be promoted by deleting
  its ``expect_block``. A stale block cannot rot silently.
"""

from __future__ import annotations

import re
import time
import traceback
from dataclasses import dataclass
from typing import Awaitable, Callable, Generic, List, Literal, Sequence, TypeVar

_C = TypeVar("_C")

SegmentOutcome = Literal["passed", "expectBlocked", "unblocked", "failed"]
"""What actually happened when a segment ran (the TS vocabulary, verbatim)."""


@dataclass(frozen=True)
class ExpectBlock:
    """Why a segment cannot pass yet, in the issue tracker's own numbers.

    Attributes:
        issues: Open issue numbers. Empty is never valid.
        because: One plain sentence: what those issues do to this segment.
        signature: The failure text those issues produce. The block excuses
            ONLY a failure matching this; anything else is a real failure.
    """

    issues: tuple[int, ...]
    because: str
    signature: "re.Pattern[str]"


@dataclass(frozen=True)
class Segment(Generic[_C]):
    """One named step of the journey.

    Attributes:
        id: The step's stable id — the step-parity unit (FR-638-022).
        title: Plain-words title, used verbatim in the printed report.
        run: Runs the real work and raises on any failed assertion.
        expect_block: Present only while a named open issue stops this
            segment passing.
    """

    id: str
    title: str
    run: Callable[[_C], Awaitable[None]]
    expect_block: ExpectBlock | None = None


@dataclass(frozen=True)
class SegmentResult:
    """One executed segment: its outcome plus the kept failure evidence."""

    id: str
    title: str
    outcome: SegmentOutcome
    duration_ms: int
    expect_block: ExpectBlock | None
    failure: str | None


@dataclass(frozen=True)
class JourneyReport:
    """Every segment result; ``ok`` iff none is ``failed`` or ``unblocked``."""

    segments: tuple[SegmentResult, ...]
    ok: bool


def _describe(error: BaseException) -> str:
    cause = error.__cause__
    suffix = "" if cause is None else f"\n  caused by: {cause!r}"
    text = f"{type(error).__name__}: {error}"
    # One frame of provenance: a journey failure is read off CI logs, and a
    # bare message with no location sends the reader grepping.
    frames = traceback.extract_tb(error.__traceback__)
    where = f" (at {frames[-1].filename}:{frames[-1].lineno})" if frames else ""
    return f"{text}{where}{suffix}"


def classify(
    expect_block: ExpectBlock | None, failure: str | None
) -> SegmentOutcome:
    """The two-direction expect-block verdict (see the module docstring)."""
    if expect_block is None:
        return "passed" if failure is None else "failed"
    if failure is None:
        return "unblocked"
    return "expectBlocked" if expect_block.signature.search(failure) else "failed"


async def run_segment(
    segment: Segment[_C],
    context: _C,
    now: Callable[[], float] = time.monotonic,
) -> SegmentResult:
    """Runs one segment and classifies it (``now`` injected for unit tests)."""
    started = now()
    failure: str | None = None
    try:
        await segment.run(context)
    except Exception as error:  # noqa: BLE001 - kept as report evidence
        failure = _describe(error)
    return SegmentResult(
        id=segment.id,
        title=segment.title,
        outcome=classify(segment.expect_block, failure),
        duration_ms=round((now() - started) * 1000),
        expect_block=segment.expect_block,
        failure=failure,
    )


@dataclass(frozen=True)
class SyncSegment(Generic[_C]):
    """One named step of the SYNC journey rewrite (Scenario 7.1).

    A separate shape rather than a union on :class:`Segment`: the sync verbs
    refuse from a thread with a running event loop (FM-638-4), so the sync
    runner below must be a plain function — typing the run callable
    ``async`` here would reintroduce the loop the rewrite exists to not have.
    """

    id: str
    title: str
    run: Callable[[_C], None]
    expect_block: ExpectBlock | None = None


def run_sync_segment(
    segment: SyncSegment[_C],
    context: _C,
    now: Callable[[], float] = time.monotonic,
) -> SegmentResult:
    """The blocking twin of :func:`run_segment`."""
    started = now()
    failure: str | None = None
    try:
        segment.run(context)
    except Exception as error:  # noqa: BLE001 - kept as report evidence
        failure = _describe(error)
    return SegmentResult(
        id=segment.id,
        title=segment.title,
        outcome=classify(segment.expect_block, failure),
        duration_ms=round((now() - started) * 1000),
        expect_block=segment.expect_block,
        failure=failure,
    )


def run_sync_journey(
    segments: Sequence[SyncSegment[_C]],
    context: _C,
    teardown: Callable[[_C], None],
) -> JourneyReport:
    """The blocking twin of :func:`run_journey` — no event loop anywhere on
    the calling thread, which is exactly the consumer shape Scenario 7.1
    reruns the narrative under."""
    results: List[SegmentResult] = []
    try:
        for segment in segments:
            results.append(run_sync_segment(segment, context))
    finally:
        teardown(context)
    return summarize(results)


async def run_journey(
    segments: Sequence[Segment[_C]],
    context: _C,
    teardown: Callable[[_C], Awaitable[None]],
) -> JourneyReport:
    """Run the segments in order, ALWAYS run ``teardown``, then summarize.

    The ``finally`` is the risky part every journey would otherwise
    hand-copy: a journey that forgets it never asks its spawned host to shut
    down (the TS kit's #24319 lesson, kept verbatim).
    """
    results: List[SegmentResult] = []
    try:
        for segment in segments:
            results.append(await run_segment(segment, context))
    finally:
        await teardown(context)
    return summarize(results)


def summarize(segments: Sequence[SegmentResult]) -> JourneyReport:
    """Folds segment results into the journey verdict."""
    return JourneyReport(
        segments=tuple(segments),
        ok=not any(s.outcome in ("failed", "unblocked") for s in segments),
    )


_MARK: dict[str, str] = {
    "passed": "PASS ",
    "expectBlocked": "BLOCK",
    "unblocked": "STALE",
    "failed": "FAIL ",
}


def _issue_list(block: ExpectBlock) -> str:
    return ", ".join(f"#{issue}" for issue in block.issues)


def _indent(text: str) -> str:
    return text.replace("\n", "\n        ")


def format_report(report: JourneyReport) -> str:
    """The human-facing report: one line per step, then the verdict line."""
    lines: List[str] = []
    for segment in report.segments:
        lines.append(
            f"{_MARK[segment.outcome]} {segment.id:<28}"
            f"{segment.duration_ms:>6}ms  {segment.title}"
        )
        block = segment.expect_block
        if block is not None and segment.outcome == "expectBlocked":
            lines.append(
                f"        expect-blocked on {_issue_list(block)}: {block.because}"
            )
            lines.append(f"        observed: {_indent(segment.failure or '(no detail)')}")
        if block is not None and segment.outcome == "unblocked":
            lines.append(
                f"        EXPECT-BLOCK IS STALE. This segment now passes, so "
                f"{_issue_list(block)} looks fixed."
            )
            lines.append(
                "        Promote it: delete this segment's expect_block so the "
                "journey requires it from now on."
            )
        if segment.outcome == "failed":
            if block is not None:
                lines.append(
                    f"        This failure does NOT match the expect-block "
                    f"signature for {_issue_list(block)} "
                    f"({block.signature.pattern}), so it is a real failure."
                )
            lines.append(f"        {_indent(segment.failure or '(no detail)')}")
    counts: dict[str, int] = {}
    for segment in report.segments:
        counts[segment.outcome] = counts.get(segment.outcome, 0) + 1
    tally = " ".join(
        f"{outcome}={counts.get(outcome, 0)}"
        for outcome in ("passed", "expectBlocked", "unblocked", "failed")
    )
    lines.append(f"journey {'OK' if report.ok else 'NOT OK'} ({tally})")
    return "\n".join(lines)
